policy: ephemeral dust followups

instagibbs opened the pull request:
Follow-up to https://github.com/bitcoin/bitcoin/pull/30239

Here are the parent PR's comments that should be addressed by this PR:

https://github.com/bitcoin/bitcoin/pull/30239/files#r1834529646
https://github.com/bitcoin/bitcoin/pull/30239/files#r1831247308
https://github.com/bitcoin/bitcoin/pull/30239/files#r1832622481
https://github.com/bitcoin/bitcoin/pull/30239/files#r1831195216
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1835805164
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1835805164
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1834639096
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1834624976
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1834619709
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1834610434
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1834504436
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1834500036
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1832985488
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1830929809
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1832376920
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1832755799
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1832492686
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1832980576
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1832784278
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1837989979
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1830996993
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1830997947
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1830012890
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1830037288
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1830977092
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1832622481
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1834726168
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1832453654
https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1848488226

instagibbs committed:
CheckEphemeralSpends: only compute txid of tx when needed

instagibbs committed:
Rename CheckValidEphemeralTx to PreCheckEphemeralTx

instagibbs committed:
Have HasDust and PreCheckValidEphemeralTx take CTransaction

instagibbs committed:
Use std::ranges for ephemeral policy checks

instagibbs committed:
Move+rename GetDustIndexes -> GetDust

Use to replace HasDust and where appropraite

instagibbs committed:
ephemeral policy: IWYU

instagibbs committed:
ephemeral policy doxygen cleanup

instagibbs committed:
bench: remove unnecessary CMTxn constructors

instagibbs committed:
unit test: assert txid returned on CheckEphemeralSpends failures

Simplify nullopt checks

instagibbs committed:
unit test: make dust index less magical

instagibbs committed:
fuzz: remove dangling reference to GetEntry

instagibbs committed:
fuzz: explain package eval coin tracking better

instagibbs committed:
fuzz: Directly place transactions in vector

instagibbs committed:
func: cleanup test_dustrelay comments

instagibbs committed:
func: cleanup reorg test comment

instagibbs committed:
fuzz: remove unused TransactionsDelta validation interface

instagibbs committed:
fuzz: use optional status instead of should_rbf_eph_spend

instagibbs committed:
fuzz: package_eval: move last_tx inside txn ctor

instagibbs committed:
unit test: adapt to changing MAX_DUST_OUTPUTS_PER_TX

instagibbs committed:
RPC: only enforce dust rules on priority when standardness active

instagibbs committed:
func: slight elaboration on submitpackage restriction

instagibbs committed:
func: rename test_free_relay to test_no_minrelay_fee

instagibbs committed:
CheckEphemeralSpends: no need to iterate inputs if no parent dust

instagibbs committed:
unit test: ephemeral_tests is using a dust relay rate, not minrelay

instagibbs committed:
func: add note about lack of 1P1C propagation in tree submitpackage

glozow commented:
Time to sweep the ephemeral top 9 commits

instagibbs committed:
CheckEphemeralSpends: return boolean, and set child state and txid outparams

instagibbs committed:
fuzz: package_eval: let fuzzer run out input in main tx creation loop

marcofleon commented:
From various fuzz runs:

<details>
<summary>stack trace 1</summary>

```
UndefinedBehaviorSanitizer:DEADLYSIGNAL
==1654160==ERROR: UndefinedBehaviorSanitizer: SEGV on unknown address 0x000000000010 (pc 0x7f4ed8ac0563 bp 0x7ffd7bdeff90 sp 0x7ffd7bdefba8 T1654160)
==1654160==The signal is caused by a READ memory access.
==1654160==Hint: address points to the zero page.
    #0 0x7f4ed8ac0563 in std::_Rb_tree_rebalance_for_erase(std::_Rb_tree_node_base*, std::_Rb_tree_node_base&) (/lib/x86_64-linux-gnu/libstdc++.so.6+0xc0563) (BuildId: 0c47cec75226c7736517d5acb61e373d541a5023)
    #1 0x5579575141b5 in std::_Rb_tree<COutPoint, COutPoint, std::_Identity<COutPoint>, std::less<COutPoint>, std::allocator<COutPoint>>::_M_erase_aux(std::_Rb_tree_const_iterator<COutPoint>) /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_tree.h:2492:26
    #2 0x5579575141b5 in std::_Rb_tree<COutPoint, COutPoint, std::_Identity<COutPoint>, std::less<COutPoint>, std::allocator<COutPoint>>::erase[abi:cxx11](std::_Rb_tree_const_iterator<COutPoint>) /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_tree.h:1197:2
    #3 0x5579575141b5 in std::set<COutPoint, std::less<COutPoint>, std::allocator<COutPoint>>::erase[abi:cxx11](std::_Rb_tree_const_iterator<COutPoint>) /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_set.h:655:21
    #4 0x5579575141b5 in (anonymous namespace)::ephemeral_package_eval_fuzz_target(std::span<unsigned char const, 18446744073709551615ul>)::$_0::operator()() const /root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/../../../../src/test/fuzz/package_eval.cpp:257:35
    #5 0x5579575141b5 in (anonymous namespace)::ephemeral_package_eval_fuzz_target(std::span<unsigned char const, 18446744073709551615ul>) /root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/../../../../src/test/fuzz/package_eval.cpp:228:27
    #6 0x5579576628c7 in std::function<void (std::span<unsigned char const, 18446744073709551615ul>)>::operator()(std::span<unsigned char const, 18446744073709551615ul>) const /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/std_function.h:591:9
    #7 0x5579576628c7 in LLVMFuzzerTestOneInput /root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/util/../../../../../src/test/fuzz/fuzz.cpp:213:5
    #8 0x5579572dfd00 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x94ed00) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #9 0x5579572df435 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x94e435) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #10 0x5579572e0bc5 in fuzzer::Fuzzer::MutateAndTestOne() (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x94fbc5) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #11 0x5579572e17c5 in fuzzer::Fuzzer::Loop(std::vector<fuzzer::SizedFile, std::allocator<fuzzer::SizedFile>>&) (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x9507c5) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #12 0x5579572cf9ef in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x93e9ef) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #13 0x5579572f8da2 in main (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x967da2) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #14 0x7f4ed8608249  (/lib/x86_64-linux-gnu/libc.so.6+0x27249) (BuildId: 58254ca972028402bc40624f81388d85ec95f70d)
    #15 0x7f4ed8608304 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x27304) (BuildId: 58254ca972028402bc40624f81388d85ec95f70d)
    #16 0x5579572c4de0 in _start (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x933de0) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
```
</details>

<details>
<summary>stack trace 2</summary>

```
../../../../src/test/fuzz/package_eval.cpp:286 operator(): Assertion `outpoints.insert(in.prevout).second' failed.
==1656011== ERROR: libFuzzer: deadly signal
    #0 0x55a766a785a4 in __sanitizer_print_stack_trace (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x9915a4) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #1 0x55a766a4e4d8 in fuzzer::PrintStackTrace() (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x9674d8) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #2 0x55a766a348f3 in fuzzer::Fuzzer::CrashCallback() (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x94d8f3) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #3 0x7f5f78a1d04f  (/lib/x86_64-linux-gnu/libc.so.6+0x3c04f) (BuildId: 58254ca972028402bc40624f81388d85ec95f70d)
    #4 0x7f5f78a6be2b  (/lib/x86_64-linux-gnu/libc.so.6+0x8ae2b) (BuildId: 58254ca972028402bc40624f81388d85ec95f70d)
    #5 0x7f5f78a1cfb1 in raise (/lib/x86_64-linux-gnu/libc.so.6+0x3bfb1) (BuildId: 58254ca972028402bc40624f81388d85ec95f70d)
    #6 0x7f5f78a07471 in abort (/lib/x86_64-linux-gnu/libc.so.6+0x26471) (BuildId: 58254ca972028402bc40624f81388d85ec95f70d)
    #7 0x55a766f77e07 in assertion_fail(std::basic_string_view<char, std::char_traits<char>>, int, std::basic_string_view<char, std::char_traits<char>>, std::basic_string_view<char, std::char_traits<char>>) /root/bitcoin/ephemeralfuzzbuild/src/util/../../../src/util/check.cpp:34:5
    #8 0x55a766c6ac7f in bool&& inline_assertion_check<true, bool>(bool&&, char const*, int, char const*, char const*) /root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/../../../../src/util/check.h:59:13
    #9 0x55a766c6ac7f in (anonymous namespace)::ephemeral_package_eval_fuzz_target(std::span<unsigned char const, 18446744073709551615ul>)::$_0::operator()() const /root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/../../../../src/test/fuzz/package_eval.cpp:286:25
    #10 0x55a766c6ac7f in (anonymous namespace)::ephemeral_package_eval_fuzz_target(std::span<unsigned char const, 18446744073709551615ul>) /root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/../../../../src/test/fuzz/package_eval.cpp:228:27
    #11 0x55a766db88c7 in std::function<void (std::span<unsigned char const, 18446744073709551615ul>)>::operator()(std::span<unsigned char const, 18446744073709551615ul>) const /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/std_function.h:591:9
    #12 0x55a766db88c7 in LLVMFuzzerTestOneInput /root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/util/../../../../../src/test/fuzz/fuzz.cpp:213:5
    #13 0x55a766a35d00 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x94ed00) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #14 0x55a766a35435 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x94e435) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #15 0x55a766a36bc5 in fuzzer::Fuzzer::MutateAndTestOne() (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x94fbc5) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #16 0x55a766a377c5 in fuzzer::Fuzzer::Loop(std::vector<fuzzer::SizedFile, std::allocator<fuzzer::SizedFile>>&) (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x9507c5) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #17 0x55a766a259ef in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x93e9ef) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #18 0x55a766a4eda2 in main (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x967da2) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #19 0x7f5f78a08249  (/lib/x86_64-linux-gnu/libc.so.6+0x27249) (BuildId: 58254ca972028402bc40624f81388d85ec95f70d)
    #20 0x7f5f78a08304 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x27304) (BuildId: 58254ca972028402bc40624f81388d85ec95f70d)
    #21 0x55a766a1ade0 in _start (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x933de0) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
 ```
</details>

<details>
<summary>stack trace 3</summary>

```
terminate called after throwing an instance of 'std::out_of_range'
  what():  map::at
==1656362== ERROR: libFuzzer: deadly signal
    #0 0x5560e2e3c5a4 in __sanitizer_print_stack_trace (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x9915a4) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #1 0x5560e2e124d8 in fuzzer::PrintStackTrace() (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x9674d8) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #2 0x5560e2df88f3 in fuzzer::Fuzzer::CrashCallback() (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x94d8f3) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #3 0x7fc27ba1d04f  (/lib/x86_64-linux-gnu/libc.so.6+0x3c04f) (BuildId: 58254ca972028402bc40624f81388d85ec95f70d)
    #4 0x7fc27ba6be2b  (/lib/x86_64-linux-gnu/libc.so.6+0x8ae2b) (BuildId: 58254ca972028402bc40624f81388d85ec95f70d)
    #5 0x7fc27ba1cfb1 in raise (/lib/x86_64-linux-gnu/libc.so.6+0x3bfb1) (BuildId: 58254ca972028402bc40624f81388d85ec95f70d)
    #6 0x7fc27ba07471 in abort (/lib/x86_64-linux-gnu/libc.so.6+0x26471) (BuildId: 58254ca972028402bc40624f81388d85ec95f70d)
    #7 0x7fc27be9d918  (/lib/x86_64-linux-gnu/libstdc++.so.6+0x9d918) (BuildId: 0c47cec75226c7736517d5acb61e373d541a5023)
    #8 0x7fc27bea8e19  (/lib/x86_64-linux-gnu/libstdc++.so.6+0xa8e19) (BuildId: 0c47cec75226c7736517d5acb61e373d541a5023)
    #9 0x7fc27bea8e84 in std::terminate() (/lib/x86_64-linux-gnu/libstdc++.so.6+0xa8e84) (BuildId: 0c47cec75226c7736517d5acb61e373d541a5023)
    #10 0x7fc27bea90d7 in __cxa_throw (/lib/x86_64-linux-gnu/libstdc++.so.6+0xa90d7) (BuildId: 0c47cec75226c7736517d5acb61e373d541a5023)
    #11 0x7fc27bea023f in std::__throw_out_of_range(char const*) (/lib/x86_64-linux-gnu/libstdc++.so.6+0xa023f) (BuildId: 0c47cec75226c7736517d5acb61e373d541a5023)
    #12 0x5560e30339be in std::map<COutPoint, long, std::less<COutPoint>, std::allocator<std::pair<COutPoint const, long>>>::at(COutPoint const&) /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_map.h:551:4
    #13 0x5560e302e1d4 in (anonymous namespace)::ephemeral_package_eval_fuzz_target(std::span<unsigned char const, 18446744073709551615ul>)::$_0::operator()() const /root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/../../../../src/test/fuzz/package_eval.cpp:260:50
    #14 0x5560e302e1d4 in (anonymous namespace)::ephemeral_package_eval_fuzz_target(std::span<unsigned char const, 18446744073709551615ul>) /root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/../../../../src/test/fuzz/package_eval.cpp:228:27
    #15 0x5560e317c8c7 in std::function<void (std::span<unsigned char const, 18446744073709551615ul>)>::operator()(std::span<unsigned char const, 18446744073709551615ul>) const /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/std_function.h:591:9
    #16 0x5560e317c8c7 in LLVMFuzzerTestOneInput /root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/util/../../../../../src/test/fuzz/fuzz.cpp:213:5
    #17 0x5560e2df9d00 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x94ed00) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #18 0x5560e2df9435 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x94e435) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #19 0x5560e2dfabc5 in fuzzer::Fuzzer::MutateAndTestOne() (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x94fbc5) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #20 0x5560e2dfb7c5 in fuzzer::Fuzzer::Loop(std::vector<fuzzer::SizedFile, std::allocator<fuzzer::SizedFile>>&) (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x9507c5) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #21 0x5560e2de99ef in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x93e9ef) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #22 0x5560e2e12da2 in main (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x967da2) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
    #23 0x7fc27ba08249  (/lib/x86_64-linux-gnu/libc.so.6+0x27249) (BuildId: 58254ca972028402bc40624f81388d85ec95f70d)
    #24 0x7fc27ba08304 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x27304) (BuildId: 58254ca972028402bc40624f81388d85ec95f70d)
    #25 0x5560e2ddede0 in _start (/root/bitcoin/ephemeralfuzzbuild/src/test/fuzz/fuzz+0x933de0) (BuildId: a5b9aaf3b4abf14131d79477c841f1529fb62bf2)
```
</details>

They all happen within a few seconds of fuzzing. The errors are non-deterministic, so the inputs aren't of much use unfortunately 😕  I'm still looking for the bug, but will likely be easier for you.

instagibbs commented:
@marcofleon to be clear this is happening on master or follow-up?

instagibbs commented:
ok replicated on this PR, I'll track it down

marcofleon commented:
It seems to be commit 7f0d8633b174c8b96cc28ec05005b8e89ab4957d

instagibbs commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -338,9 +333,6 @@ FUZZ_TARGET(ephemeral_package_eval, .init = initialize_tx_pool)
-
         CheckMempoolEphemeralInvariants(tx_pool);
     }
 
```
specifically needs the `SyncWithValidationInterfaceQueue`, not the unused part.

naumenkogs commented on src/policy/ephemeral_policy.h:
```diff
@@ -50,8 +50,9 @@ bool PreCheckValidEphemeralTx(const CTransaction& tx, CFeeRate dust_relay_rate,
  *  Checks that each transaction's parents have their dust spent by the child,
  *  where parents are either in the mempool or in the package itself.
- *  @returns std::nullopt if all dust is properly spent, or the txid of the violating child spend.
+ *  child_state and child_txid set on failure.
```
f306a940d2ba470a13120853ace49509a8e496a6

looks odd, perhaps a pasting issue.

instagibbs replied:
It's a sentence fragment, I can touch it up to be grammatically correct

instagibbs replied:
touched up

instagibbs commented on src/test/transaction_tests.cpp:
```diff
@@ -814,9 +814,11 @@ BOOST_AUTO_TEST_CASE(test_IsStandard)
-    // Add dust output to take dust slot, still standard!
-    t.vout.emplace_back(0, t.vout[0].scriptPubKey);
-    CheckIsStandard(t);
+    // Add dust outputs up to allowed maximum, still standard!
```
we do in the following lines

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -10,7 +10,7 @@ bool HasDust(const CTransactionRef& tx, CFeeRate dust_relay_rate)
 }
 
-bool CheckValidEphemeralTx(const CTransactionRef& tx, CFeeRate dust_relay_rate, CAmount base_fee, CAmount mod_fee, TxValidationState& state)
+bool PreCheckValidEphemeralTx(const CTransactionRef& tx, CFeeRate dust_relay_rate, CAmount base_fee, CAmount mod_fee, TxValidationState& state)
```
done

naumenkogs commented:
ACK 5679f398e2a3ab8315d39ee674ccb8ad6a8f73c7

instagibbs committed:
functional: only generate required blocks for test

hodlinator commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -210,50 +210,47 @@ FUZZ_TARGET(ephemeral_package_eval, .init = initialize_tx_pool)
+                // Note that this test currently only spends package outputs in last transaction.
+                bool last_tx = num_txs > 1 && txs.size() == num_txs - 1;
+                const auto num_in = outpoint_to_rbf ? 2 :
                     last_tx ? fuzzed_data_provider.ConsumeIntegralInRange<int>(package_outpoints.size()/2 + 1, package_outpoints.size()) :
                     fuzzed_data_provider.ConsumeIntegralInRange<int>(1, 4);
-                auto num_out = should_rbf_eph_spend ? 1 : fuzzed_data_provider.ConsumeIntegralInRange<int>(1, 4);
+                auto num_out = outpoint_to_rbf ? 1 : fuzzed_data_provider.ConsumeIntegralInRange<int>(1, 4);
```
nit: Could add `const` to `num_out` for symmetry with `num_in` as you are touching this line (pointed out in https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1832788903).

instagibbs replied:
done

hodlinator commented on src/test/txvalidation_tests.cpp:
```diff
@@ -128,85 +134,145 @@ BOOST_FIXTURE_TEST_CASE(ephemeral_tests, RegTestingSetup)
-    BOOST_CHECK(CheckEphemeralSpends({grandparent_tx_1, dust_spend, dust_non_spend}, minrelay, pool).has_value());
-    BOOST_CHECK(CheckEphemeralSpends({grandparent_tx_1, dust_non_spend}, minrelay, pool).has_value());
+    const auto dust_non_spend_txid{dust_non_spend->GetHash()};
+    const Txid null_txid;
+    assert(dust_non_spend_txid != null_txid);
```
In 0f92fa2d468186a1526827a063f66e3042e028dc:
No longer used, should probably be removed there.

instagibbs replied:
removed

hodlinator commented on src/policy/policy.cpp:
```diff
@@ -142,13 +150,11 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
         }
     }
 
     // Only MAX_DUST_OUTPUTS_PER_TX dust is permitted(on otherwise valid ephemeral dust)
```
nit:
```suggestion
    // Only MAX_DUST_OUTPUTS_PER_TX dust is permitted (on otherwise valid ephemeral dust)
```

instagibbs replied:
leaving as is

hodlinator reviewed (changes requested):
Reviewed 5679f398e2a3ab8315d39ee674ccb8ad6a8f73c7

Especially like the refactor of `CheckEphemeralSpends()` with less surprising return/out-values.

Good to see `LIMITED_WHILE(fuzzed_data_provider.remaining_bytes() > 0` in more places than I suggested.

Suggested changes aside from inline comments:

### 03ec3f6230f440b264abe3aa42a57fdd29bd3ba3 commit message

```diff
- Implement GetDust to replace HasDust, use where appropriate
+ Move+rename GetDustIndexes -> GetDust to replace HasDust, use where appropriate
```

### Deduplicate `assert_mempool_contents()`

As [pointed out by theStack & me](https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1825278134):
```diff
diff --git a/test/functional/mempool_package_rbf.py b/test/functional/mempool_package_rbf.py
index a5b8fa5f87..14194f9e1a 100755
--- a/test/functional/mempool_package_rbf.py
+++ b/test/functional/mempool_package_rbf.py
@@ -19,6 +19,7 @@ from test_framework.wallet import (
     DEFAULT_FEE,
     MiniWallet,
 )
+from test_framework import mempool_util
 
 MAX_REPLACEMENT_CANDIDATES = 100
 
@@ -37,15 +38,7 @@ class PackageRBFTest(BitcoinTestFramework):
         ]] * self.num_nodes
 
     def assert_mempool_contents(self, expected=None):
-        """Assert that all transactions in expected are in the mempool,
-        and no additional ones exist.
-        """
-        if not expected:
-            expected = []
-        mempool = self.nodes[0].getrawmempool(verbose=False)
-        assert_equal(len(mempool), len(expected))
-        for tx in expected:
-            assert tx.rehash() in mempool
+        mempool_util.assert_mempool_contents(self, self.nodes[0], expected, sync=False)
 
     def create_simple_package(self, parent_coin, parent_fee=DEFAULT_FEE, child_fee=DEFAULT_CHILD_FEE, heavy_child=False):
         """Create a 1 parent 1 child package using the coin passed in as the parent's input. The
```

### + Sanity check for `assert_mempool_contents()`

Could add sanity check if passing same tx twice in `expected` (https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1831120099)
```diff
diff --git a/test/functional/test_framework/mempool_util.py b/test/functional/test_framework/mempool_util.py
index e04ae914cc..72da52d957 100644
--- a/test/functional/test_framework/mempool_util.py
+++ b/test/functional/test_framework/mempool_util.py
@@ -30,6 +30,9 @@ def assert_mempool_contents(test_framework, node, expected=None, sync=True):
         test_framework.sync_mempools()
     if not expected:
         expected = []
+    else:
+        # Sanity check that elements are unique.
+        assert_equal(len(expected), len(set(expected)))
     mempool = node.getrawmempool(verbose=False)
     assert_equal(len(mempool), len(expected))
     for tx in expected:
```

instagibbs committed:
assert_mempool_contents: assert not duplicates expected

instagibbs commented on src/test/txvalidation_tests.cpp:
```diff
@@ -100,9 +100,11 @@ static inline CTransactionRef make_ephemeral_tx(const std::vector<COutPoint>& in
     }
-    for (auto i{0}; i < 3; ++i) {
+    int num_outputs{3};
+    int dust_index{num_outputs - 1};
```
took modified version of file-local constant

instagibbs commented on src/test/txvalidation_tests.cpp:
```diff
@@ -128,85 +128,88 @@ BOOST_FIXTURE_TEST_CASE(ephemeral_tests, RegTestingSetup)
+    const auto dust_non_spend_txid{dust_non_spend->GetHash()};
+    const Txid null_txid;
+    assert(dust_non_spend_txid != null_txid);
+    BOOST_CHECK_EQUAL(CheckEphemeralSpends({grandparent_tx_1, dust_non_spend, dust_spend}, minrelay, pool).value_or(null_txid), dust_non_spend_txid);
```
IIUC I'd have needed to at least make the RHS an optional to do that?

Either way, this goes away in a later commit, so I'm going to leave as-s.

instagibbs commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -210,50 +210,47 @@ FUZZ_TARGET(ephemeral_package_eval, .init = initialize_tx_pool)
-
             // Create transaction to add to the mempool
-            const CTransactionRef tx = [&] {
+            txs.push_back([&] {
```
done

instagibbs commented on src/rpc/mining.cpp:
```diff
@@ -496,7 +496,7 @@ static RPCHelpMan prioritisetransaction()
     // Non-0 fee dust transactions are not allowed for entry, and modification not allowed afterwards
     const auto& tx = mempool.get(hash);
-    if (tx && HasDust(tx, mempool.m_opts.dust_relay_feerate)) {
+    if (mempool.m_opts.require_standard && tx && !GetDust(*tx, mempool.m_opts.dust_relay_feerate).empty()) {
```
Setting `-dustrelayfee=0` avoids the check much more directly and simply, with no recompilation needed.

Gating it based on standardness checks is just easier to understand when e.g., standardness isn't being enforced anyways, so dust should be treated as a no-op.

I'm going to mark this as resolved. If we want to reverse this logic, imo we should do it in another PR that removes both the logic in prioritisation as well as the modified fee checks.

glozow replied:
fwiw I'm not too concerned about this "encouraging" miners to patch. Miners might mine dust and this doesn't stop them from doing anything, but I don't think this would make that more likely given "subverting" it would only require `-dustrelayfee=0` and has always been possible. I think it's worth discussing and asking miners. But that's my thinking.

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -63,16 +72,23 @@ std::optional<Txid> CheckEphemeralSpends(const Package& package, CFeeRate dust_r
             processed_parent_set.insert(parent_txid);
```
going to leave as is

instagibbs commented:
reworked https://github.com/bitcoin/bitcoin/commit/03ec3f6230f440b264abe3aa42a57fdd29bd3ba3 commit message

> Deduplicate assert_mempool_contents()
Deferring to future work since that's a different test, I'll review it if someone opens a clenaup PR.

>  Sanity check for assert_mempool_contents()

Good idea, added minimal diff

naumenkogs commented:
ACK 466e4df3fb83ef82b6add22e202e3a70dbf83a12

a trivial refactoring since the last ack

hodlinator reviewed (approved):
ACK 466e4df3fb83ef82b6add22e202e3a70dbf83a12

Mostly responded to my feedback since my last review.

Compiled with BDB and passed functional tests, unittests and ran `ephemeral_package_eval` fuzz target for 1 minute.

Still have [reservations about the prioritisetransaction RPC](https://github.com/bitcoin/bitcoin/pull/31279/files#r1848453460), but not blocking.

theStack reviewed (approved):
lgtm ACK 466e4df3fb83ef82b6add22e202e3a70dbf83a12

If you need to retouch, could also tackle https://github.com/bitcoin/bitcoin/pull/30239#discussion_r1828577696 (e.g. put it into c5c10fd317c6b4c033f3001757e6975b8b9a4942)

glozow commented on src/policy/policy.h:
```diff
@@ -131,6 +131,8 @@ bool IsDust(const CTxOut& txout, const CFeeRate& dustRelayFee);
 bool IsStandard(const CScript& scriptPubKey, const std::optional<unsigned>& max_datacarrier_bytes, TxoutType& whichType);
 
+/** Get the vout index numbers of all dust outputs */
+std::vector<uint32_t> GetDust(const CTransaction& tx, CFeeRate dust_relay_rate);
```
Why not `CTransactionRef`?

glozow commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -5,12 +5,12 @@
 #include <policy/policy.h>
 
-bool HasDust(const CTransactionRef& tx, CFeeRate dust_relay_rate)
+bool HasDust(const CTransaction& tx, CFeeRate dust_relay_rate)
```
Why not `CTransactionRef`? (3ed930a1f41f7d7160c6ede5dcf3d4d5f1cfa876)

glozow commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -210,7 +210,7 @@ FUZZ_TARGET(ephemeral_package_eval, .init = initialize_tx_pool)
     chainstate.SetMempool(&tx_pool);
 
-    LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 300)
+    LIMITED_WHILE(fuzzed_data_provider.remaining_bytes() > 0, 300)
```
oh edit, i see now
![image](https://github.com/user-attachments/assets/e8eb149d-98ae-4dc8-a05f-7bd7b8b4339e)

glozow reviewed (commented):
utACK 466e4df3fb83ef82b6add22e202e3a70dbf83a12
